eclipse-theia/theia
# Resolve latest version of `ms-vscode' builtins

#### What it does

The base ovsx-client recognizes `builtins` by the namespace
e.g. `vscode` and `theia` before resolving the latest compatible extension,
however there are additional builtins which can not be identified by the namespace.

E.g.
Namespace: 'ms-vscode'
Extension names: 'node-debug', 'node-debug2','references-view', 'js-debug-companion',
'js-debug'.

This change includes the extensions above in the mechanism that
selects the latest compatible version for builtins.

#### How to test


1) remove the plugins folder and temp files (ex: rm -rf plugins/, rm -rf /tmp/vscode-unpacked, rm -rf ~./theia/extensions).
2) Remove builtin `ms-vscode` etensions from the root package.json file i.e.
    "vscode-builtin-js-debug": "https://open-vsx.org/api/ms-vscode/js-debug/1.49.8/file/ms-vscode.js-debug-1.49.8.vsix",
    "vscode-builtin-node-debug": "https://open-vsx.org/api/ms-vscode/node-debug/1.44.8/file/ms-vscode.node-debug-1.44.8.vsix",
    "vscode-builtin-node-debug2": "https://open-vsx.org/api/ms-vscode/node-debug2/1.42.1/file/ms-vscode.node-debug2-1.42.1.vsix",
    "vscode-references-view": "https://open-vsx.org/api/ms-vscode/references-view/0.0.47/file/ms-vscode.references-view-0.0.47.vsix"

3) start the application.
4) search for `ms-vscode` builtin extensions (ex: js-debug - the extension should be at 1.50.0 (current API level).
5) search for `ms-vscode` builtin extensions (ex: js-debug-companion - the extension should be at 1.0.13 (latest below API level).
![js-debugger-after](https://user-images.githubusercontent.com/76971376/122127968-f4115d80-ce01-11eb-8665-e8162f59649c.png)

6) searching other extensions should behave like today.
7) installing extensions should work correctly.
8)  repeat step 1 - ensure that no builtins are present.
9) Download the following extension package file, change the extension from .zip to .vsix
[ms-vscode.zip](https://github.com/eclipse-theia/theia/files/6658996/ms-vscode.zip)
10) Install the .vsix extension package via the extensions-view option "install from VSIX"
11) ensure that extensions in the pack are resolved to version 1.50.0 if it exists, else a previous version if it exists.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
